src/update_notifier.rs

```diff
@@ -3,117 +3,231 @@
 //! This is turned on in our releases by default but is off when building from source
 
 use crate::{config, proxy, util};
+use chrono::{DateTime, Utc};
 use colored::*;
 use reqwest::{
     header::{HeaderValue, ACCEPT},
     Client, RedirectPolicy, Response,
 };
 use std::fmt::Write;
+use std::fs::File;
+use std::path::PathBuf;
 
 const GITHUB_RELEASE_PAGE: &str = "https://github.com/wasmerio/wasmer/releases/latest";
 const GITHUB_RELEASE_URL_BASE: &str = "https://github.com/wasmerio/wasmer/releases/tag/";
+const GLOBAL_WAPM_UPDATE_FILE: &str = ".wapm_update.json";
+const BACKGROUND_UPDATE_CHECK_RUNNING: &str = ".background_update_process_running.txt";
+
+/// The amount of seconds that we need to wait until showing the next notification
+const CHECK_DURATION_IN_SECONDS: u64 = 60 * 60 * 24; // 24 hours
+const WAPM_NOTIFICATION_WINDOW: u64 = 60 * 60 * 2; // 2 hours
 
 #[derive(Debug, Deserialize)]
 struct VersionResponse {
     tag_name: String,
 }
 
-/// this is the base call, it will spawn another process
-pub fn run_async_check_base() -> Option<()> {
-    if let Some((last_checked_time, maybe_next_version)) =
-        config::Config::update_notifications_enabled()
-            .and_then(|()| config::get_last_update_checked_time())
-    {
-        // if we have it cached, then call check to pull it out of the cache and print it
-        if let Some(message) = maybe_next_version
-            .and_then(|next_version| check(Some(next_version)))
-            .and_then(|res| format_message(&res.old_version, &res.new_version, &res.release_url))
-        {
-            print_message(&message);
-            // clear the cache
-            config::set_last_update_checked_time(None);
-        } else {
-            // otherwise, if it's time to check again, spawn a background process to update the cache
-            let now = time::now();
-            let time_to_check: time::Duration =
-                time::Duration::from_std(std::time::Duration::from_secs(60 * 60 * 24)).unwrap();
-            if now - last_checked_time >= time_to_check {
-                config::lock_background_process()?;
-                // lock and check for lock
-                std::process::Command::new("wapm")
-                    .arg("run-background-update-check")
-                    .stdout(std::process::Stdio::null())
-                    .stderr(std::process::Stdio::null())
-                    .spawn()
-                    .ok()?;
+#[derive(Deserialize, Serialize, Debug, PartialEq)]
+pub struct WapmUpdate {
+    /// The data related to the last check on the Github Registry
+    pub last_check: Option<WapmLastCheck>,
+    /// The time when wapm last trigger the notification
+    pub last_notified: Option<DateTime<Utc>>,
+}
+
+impl WapmUpdate {
+    fn default() -> Self {
```

Comment:
⛳️ instead add `derive(Default)` you can still call `WapmUpdate::default()` and it'll do exactly the same thing